### PR TITLE
Feature/20540 prevent duplicate answers

### PR DESCRIPTION
**Pull request by jon-shipley**
Prevent the possibility of accepting duplicate answers to the same question.  Instead, and new audit is recorded with the details of the duplicate.

### CODE REVIEW

File: pupil-spa/src/app/services/answer/answer.service.ts [hunk 1]
```diff
@@ -15,8 +18,24 @@ export class AnswerService {
...
+      x => {
+        if (x.sequenceNumber === answer.sequenceNumber
+        && x.factor1 === answer.factor1
+        && x.factor2 === answer.factor2) {
+          return true;
+        }
+        return false;
+      });
+
+    if (isAlreadyAnswered.length) {
```
Comment (GuyHarwood) [inline pupil-spa/src/app/services/answer/answer.service.ts]: will this not always be initialised to zero or more?

Comment (jon-shipley) [reply pupil-spa/src/app/services/answer/answer.service.ts]: It's the result of a `filter` so will be an array.  If it finds any there will an element in the array.  Perhaps the name suggests it is a boolean when it isn't.  What do you suggest?

File: pupil-spa/src/app/services/answer/answer.service.ts [hunk 2]
```diff
@@ -15,8 +18,24 @@ export class AnswerService {
       answers = [];
     }
 
+    // Check for duplicates - we can only store one answer for each question
+    const isAlreadyAnswered = answers.filter(
+      x => {
+        if (x.sequenceNumber === answer.sequenceNumber
```
Comment (GeorgeChatzigiannis) [inline pupil-spa/src/app/services/answer/answer.service.ts]: why not `x => x.sequenceNumber === answer.sequenceNumber && x.factor1 === answer.factor1 && x.factor2 === answer.factor2` ?

File: pupil-spa/src/app/services/answer/answer.service.ts (hunk 1 above)
Comment (danwilliams) [reply pupil-spa/src/app/services/answer/answer.service.ts]: I agree that this sounds like a boolean. `alreadyAnswered` would work nicely.

**Comment by jon-shipley** [comment]
@dan - Thanks for the comment.

Changes have already been added in commit https://github.com/DFEAGILEDEVOPS/MTC/pull/662/commits/8f5721cbcd5d0d9ca7f93839b1e4a1ea6e627544

```
+    if (previousAnswers.length > 0) {
 +      // Log this error
 +      this.auditService.addEntry(new DuplicateAnswerError(answer));
 +      return;
 +    }
```

File: pupil-spa/src/app/services/answer/answer.service.ts [hunk 3]
```diff
@@ -15,8 +18,33 @@ export class AnswerService {
...
 
+  /**
+   * Find a previous answer in the answers array
+   * @param {Answer} answer - the current answer to be stored
+   * @param [Answer] answers - the list of previously stored answers
+   * @return [Answer?] - empty array or array of answers
+   */
+  private getPreviousAnswers(answer: Answer, answers) {
+    return answers.filter(
+      x => {
```
Comment (GeorgeChatzigiannis) [inline pupil-spa/src/app/services/answer/answer.service.ts]: Isn't it more readable like this? `x => x.sequenceNumber === answer.sequenceNumber && x.factor1 === answer.factor1 && x.factor2 === answer.factor2`
You are basically using an if to return boolean

Comment (jon-shipley) [reply pupil-spa/src/app/services/answer/answer.service.ts]: @GeorgeChatzigiannis - I thought it was less readable that way, sorry.

Comment (GeorgeChatzigiannis) [reply pupil-spa/src/app/services/answer/answer.service.ts]: I guess readability can be subjective

Comment (jon-shipley) [reply pupil-spa/src/app/services/answer/answer.service.ts]: @GeorgeChatzigiannis - happy to oblige if there is sufficient support for the change

Comment (GeorgeChatzigiannis) [reply pupil-spa/src/app/services/answer/answer.service.ts]: @jon-shipley it's a minor detail really if everyone else approves it as it stands I am happy.

**Pull request merged**